### PR TITLE
fix a bug in upper bound of kernels that loop along tiles in a row

### PR DESCRIPTION
**Pull request by iclementine**
Fix a bug in upper bound of kernels that loop along tiles in a row(N-dimension) to avoid illegal memory access. (`_fwd_kernel` and `_bwd_kv_kernel`). This bug only causes illegal memory access when `N` is divisible by `BLOCK_N`.

`N_CTX + P_SEQ` is the sequence length of k & v. `N_CTX` is the sequence length of Q.

before 

```python
if IS_CAUSAL:
	hi = P_SEQ + (start_m + 1) * BLOCK_M
else:
	hi = N_CTX + P_SEQ
```

after 

```python
if IS_CAUSAL:
	hi = tl.minimum(N_CTX + P_SEQ, P_SEQ + (start_m + 1) * BLOCK_M)
else:
	hi = N_CTX + P_SEQ
```

In the loop, k/v is loaded without masking on dimension N when N is divisible by BLOCK_N.

```python
    # loop over a row
    for start_n in range(0, hi, BLOCK_N):
        offs_n = start_n + offs_n_base
        
        # load k1, k2, v on chip
        if DIVISIBLE_N:
            v = tl.load(v_ptrs)
            k = tl.load(k_ptrs)
        else:
            mask_n = offs_n < (N_CTX + P_SEQ)
            v = tl.load(v_ptrs, mask=mask_n[:, None])
            k = tl.load(k_ptrs, mask=mask_n[:, None])
```

Actually, `(start_m + 1) * BLOCK_M` is the upper bound of M-positions of this CTA. So in the sense of causal masking, the upper bound of N-position of this CTA is `P_SEQ + (start_m + 1) * BLOCK_M`. However, merely using `P_SEQ + (start_m + 1) * BLOCK_M` as the upper bound is problematic since it may exceeds the sequence length of k/v (`P_SEQ + N_CTX`). So we should further resrict that it would not exceeds `P_SEQ + N_CTX`. 

```python
hi = tl.minimum(N_CTX + P_SEQ, P_SEQ + (start_m + 1) * BLOCK_M)
```

Our implementation has this issue because we support `M !=N`(actually only `N >=M` is considered carefully. `P_SEQ` here means previous sequence length), which introduces a lot of details to take care of. But after all, we can visualize it and fix it.

# Causal Masking

In the following 2 cases, the rule for causal masking is consistent. The rectangle with solid line means the area of computatioin.

The area below the -45° diagonal line across the upper-left vertex of `Q*K` area is "causal".
图片

# Valid Masking

Since we use a CTA-level representation of program & TILE-level representation of data in triton language. When M or N is not divisible by the corresponding tile size along that dimension, masking is required, As shown below.

图片

# Combine them together
Consider valid masking & tile level representation of data. We need to specify the loop bound in a safe and proper way. Most of the time specifying how many tiles to loop over is more intuitive that the loop bound of M-position or N-position, though we can transform them back and forth via some integer arithmetics. Here is the visualization of why `hi=(start_m +1) * BLOCK_M + P_SEQ` is problematic.

图片

In the cases above `hi=(start_m +1) * BLOCK_M + P_SEQ` would exceed the sequence length of k/v. Then If we do not apply masking when loading k/v, there would be illegal memory access.

So we use `hi=tl.minumum(P_SEQ+N_CTX, (start_m +1) * BLOCK_M + P_SEQ)` instead.

### CODE REVIEW

File: src/flag_attn/flash.py [hunk 1]
```diff
@@ -289,8 +289,13 @@ def _fwd_kernel(
...
     #     q = tl.dot(I, q).to(input_dtype)
 
+    # NOTE: Loop-Bound-For-N
+    # The upper bound of index in n should never exceed the sequence length of k/v. 
+    # `hi = P_SEQ + (start_m + 1) * BLOCK_M` alone does not consider the valid sequence 
+    # length of k/v. It may cause illegal memory access when loading k & v tiles 
+    # if mask_n is not applied for loading.
     if IS_CAUSAL:
-        hi = P_SEQ + (start_m + 1) * BLOCK_M
+        hi = tl.minimum(N_CTX + P_SEQ, P_SEQ + (start_m + 1) * BLOCK_M)
```
[the remark this answers is not in the document]
Comment (tongxin) [reply src/flag_attn/flash.py:298]: I would've rewritten this piece to the following to cut one multiplication
```
...
# in line 259, make a value for later reuse
m_start = start_m * BLOCK_M
offs_m = m_start + offs_m_base 
...
hi = N_CTX + P_SEQ
if IS_CAUSAL:
    hi = tl.minimum(hi, m_start + BLOCK_M)
...

**Review by tongxin — APPROVED** [review]
o/w looks good!

File: src/flag_attn/flash.py [hunk 2]
```diff
@@ -289,8 +289,13 @@ def _fwd_kernel(
     #                  tl.full((BLOCK_M, BLOCK_M), 0.0, dtype=input_dtype))
     #     q = tl.dot(I, q).to(input_dtype)
 
+    # NOTE: Loop-Bound-For-N
+    # The upper bound of index in n should never exceed the sequence length of k/v. 
+    # `hi = P_SEQ + (start_m + 1) * BLOCK_M` alone does not consider the valid sequence 
+    # length of k/v. It may cause illegal memory access when loading k & v tiles 
+    # if mask_n is not applied for loading.
```
[the remark this answers is not in the document]
Comment (iclementine) [reply src/flag_attn/flash.py:296]: More detailed explanation and a link to this page is added.

File: src/flag_attn/flash.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (iclementine) [reply src/flag_attn/flash.py:302]: This computation is once-for-a-CTX. Maybe the difference it not evident.

**Pull request merged**